### PR TITLE
Add context to Start, Init, Stop, and Finalize methods

### LINKED ISSUE #5: Provide a way to use a context when running a process
In a recent code review with my team one of the questions they brought up was if there was a way to provide a "base context" to each process while it's running. I've thought about it myself at times and I think it would be beneficial to have a context passed in to the `Start()` method on a process. Maybe the `Init()` method too, but I think having an "init context" makes a little less sense than having a "runtime context". I've been trying to think of a way to make the context available in the `Start()` method in a backward-compatible way but so far I haven't actually come up with an idea I like so far and wanted to see what you thought.

The best option, I think, would be to just make `Start() error` into `Start(ctx context.Context) error`, but that obviously wouldn't be backward compatible. Would it really be so bad to make that as a breaking change, though? It's a simple update and maybe Nacelle's ready for v2? :grin: 

The next option I thought of was a `StartWithContext(ctx context.Context) error`, but that runs into the same issue we had when talking about #4, where it modifies the fundamental `Process` interface into an either/or, so I'm pretty sure that's a no-go.

Another idea I had was to do something similar to what we talked about with `Inject(ServiceContainer) error` in #4 would be to create something like a `InjectContext(ctx context.Context)`, or a `BaseContext(ctx context.Context)`. It would feel similar in design to `BaseContext` on `http.Server` to me for setting a base request context, but I have a problem with it not being passed in to the method the context applies to.

So yeah, I'm not sure of a good way to expose that to the process itself. As far as the bootstrap side, I thought there could be both a bootstrap config level and a process level config to set a base context:

```go
ctx := context.Background()

func setupFunc(processes nacelle.ProcessContainer, services nacelle.ServiceContainer) error {
    processes.RegisterProcess(foo.New()) // Uses BaseContext from bootstrapper
    processes.RegisterProcess(bar.New(), nacelle.WithContext(ctx)) // Uses context passed in
}

nacelle.NewBootstrapper("asdf", setupFunc, nacelle.WithBaseContext(ctx))
```

### PR DESCRIPTION
**Pull request by efritz**
Closes https://github.com/go-nacelle/nacelle/issues/5.

Changes:

- Start, Init, Stop, and Finalize methods now accept a context (supplied via Runner.Run)
- Made Process optionally conform to Finalize interface
- Context for Start is canceled after calling Stop (but before calling Finalize)

### CODE REVIEW

**Review by aphistic — COMMENTED** [review]
Would this be the appropriate PR to add Process and Init options for `ProcessBaseContext` and `InitializerBaseContext`? I envision those working similar to how `http.Server.BaseContext` works where you pass in a function of `func(context.Context) context.Context` to allow you to modify the context derived for each of the `Process`'s functions. In addition to both `ProcessBaseContext` and `InitializerBaseContext` I thought that the bootstrap could have a `WithBaseContext` with a similar signature that would be run before all those other individual methods run. This way a user could choose to either have an overall modification for all the contexts, or just modify the one for their specific process or initializer.

File: parallel_initializer.go [hunk 1]
```diff
@@ -163,9 +164,9 @@ func (pi *ParallelInitializer) finalizeAll(initErrors []error) []error {
 	return errors
 }
 
-func (pi *ParallelInitializer) initWithTimeout(initializer namedInitializer, config config.Config) error {
+func (pi *ParallelInitializer) initWithTimeout(ctx context.Context, initializer namedInitializer, config config.Config) error {
```
Comment (aphistic) [inline parallel_initializer.go:167]: Could this and the `finalizeWithTimeout` take advantage of using `context.WithTimeout` at a higher level to get rid of the timeout-specific code here?

**Comment by aphistic** [comment]
We may also want to call out in the docs specifically which domains a context is associated with. If you have a timeout/deadline/etc set on a context, is that same context used for `Init`, `Start` and `Stop`, or is it applied to each of those individually?

File: parallel_initializer.go (hunk 1 above)
Comment (efritz) [reply parallel_initializer.go:167]: I'm not doing any cleanup yet, this is just to get the public interface into a state that works as expected. I'm planning to do a cleanup refactor pass afterwards.

**Comment by efritz** [comment]
Docs will be out of date as they're transferred to the new docsite and rewritten.

**Comment by efritz** [comment]
@aphistic Updated. Will still need to add a WithContext/WithContextFilter for the context passed to Run from the boostrapper.

**Comment by aphistic** [comment]
I started writing a comment awhile ago but I guess I never actually sent it. =/ I'm not sure I like the term "Filter" for the thing you use to modify the context because, to me, it implies you're removing something when it's actually the opposite in this case.

**Comment by efritz** [comment]
> I started writing a comment awhile ago but I guess I never actually sent it. =/ I'm not sure I like the term "Filter" for the thing you use to modify the context because, to me, it implies you're removing something when it's actually the opposite in this case.

I'm not married to those names. I can update them globally before I tag v2.

Could you possibly make a list of other clunky names I could take care of around the same time?

**Comment by aphistic** [comment]
> Could you possibly make a list of other clunky names I could take care of around the same time?

Sure, do you know when you'll need that? I can make sure to have it reviewed by then.

**Comment by efritz** [comment]
Anytime before we cut the v2 branch since I'd want to get rid of the bad ones immediately.

**Comment by efritz** [comment]
Just mention them as you see them and I'll do a quick rename everywhere.

**Pull request merged**